### PR TITLE
shared control plane multicluster fixes

### PR DESCRIPTION
**Pull request by ayj**
* Rename the remote istiod service and endpoint to `istiod-remote` to
  avoid conflicts with the real local istiod service.

* Use the `istiod-remote.<namespace>.svc` hostname for the sidecar and
  ingress proxies discoveryAddress. This address needs to match the
  SAN in istiod's cert. The `istiod-remote` headless service will
  resolve the hostname to the remote IP address.

* Add the `istiod-remote` hostname to istiod's SANs. Also use istiod's
  namespace to construct the legacy service names instead of
  hardcoding them to `istio-system`.

* Simplify the remote profile by removing redundant and unused values.

Usage:

* master
```bash
kubeconfig config use-context=<master>
kubectl create namespace istio-system
kubectl create secret generic cacerts -n istio-system \
    --from-file=samples/certs/ca-cert.pem \
    --from-file=samples/certs/ca-key.pem \
    --from-file=samples/certs/root-cert.pem \
    --from-file=samples/certs/cert-chain.pem

# Option (1) - expose istiod through the istio-ingressgateway
istioctl manifest apply --set values.global.meshExpansion.enabled=true
# wait for ingress external IP to be ready
ISTIOD_REMOTE=$(kubectl -n istio-system get svc istio-ingressgateway -o jsonpath='{.status.loadBalancer.ingress[0].ip}')

# Option (2) - create a dedicated control plane gateway 
istioctl manifest apply 
istioctl manifest apply -f <dedicated-control-plane-gw>
ISTIOD_REMOTE=<dedicated gateway external IP>

# Option (3)
istioctl manifest apply -f <profile-with-istiod-ilb>
ISTIOD_REMOTE=$(kubectl -n istio-system get svc istiod -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
```

* remote
```bash
kubeconfig config use-context=<remote>
kubectl create namespace istio-system
kubectl create secret generic cacerts -n istio-system \
    --from-file=samples/certs/ca-cert.pem \
    --from-file=samples/certs/ca-key.pem \
    --from-file=samples/certs/root-cert.pem \
    --from-file=samples/certs/cert-chain.pem
istioctl manifest apply --set values.global.remotePilotAddress=$(ISTIOD_REMOTE)
istioctl x create-remote-secret | kubectl --context=<master> apply -f -
```

### CODE REVIEW

**Comment by ayj** [comment]
Keeping this as a draft while I verify the cleaned up changes against the revised docs (PR pending).

File: manifests/base/templates/endpoints.yaml
Comment (linsun) [inline manifests/base/templates/endpoints.yaml]: is istio-pilot endpoint really needed?

File: manifests/base/templates/services.yaml
Comment (linsun) [inline manifests/base/templates/services.yaml]: Q - do we purposefully not want selector for this service?

File: manifests/base/templates/endpoints.yaml
Comment (ayj) [reply manifests/base/templates/endpoints.yaml]: It would only be needed if we need to support multicluster with the pre-istiod deployment model.

File: manifests/base/templates/services.yaml
Comment (ayj) [reply manifests/base/templates/services.yaml]: Yes, its a headless / selector-less service (see https://kubernetes.io/docs/concepts/services-networking/service/#without-selectors). It isn't backed by a pod in this cluster.

File: operator/data/profiles/remote.yaml
Comment (ayj) [inline operator/data/profiles/remote.yaml]: I think we should remove this and remote should inherit from the default profile. If Prometheus is a separate setup step for the master cluster it should also be a separate step in the remote. @mandarjog @costinm, wdyt? With this change we could remove the remote profile.

**Comment by linsun** [comment]
@ayj thank you for the PR!  Looks reasonable - definitely I run into some of the challenges you fix in the PR as well.  Having it changed to istiod-remote for xDS serving reduces a lot of confusing.

File: manifests/base/templates/endpoints.yaml
Comment (linsun) [reply manifests/base/templates/endpoints.yaml]: ok SG

Comment (linsun) [reply manifests/base/templates/endpoints.yaml]: BTW - I think the istiod & istio-pilot endpoints will also be created as the result of pilot.enabled=true.  so do we still need to create this again?

File: operator/data/profiles/remote.yaml
Comment (linsun) [reply operator/data/profiles/remote.yaml]: don't default profile has istioRempte=false?  also, what about the network configuration for network 1 and network 2?

File: manifests/base/templates/endpoints.yaml
Comment (ayj) [reply manifests/base/templates/endpoints.yaml]: `istio-pilot` shouldn't be needed in 1.6+. I kept it because it would make backporting to 1.5 a bit easier. If someone was using the legacy charts and separate services in 1.5 they wouldn't be enabling pilot in the remote cluster.

File: operator/data/profiles/remote.yaml
Comment (ayj) [reply operator/data/profiles/remote.yaml]: I don't think we need istioRemote. Checking remote{Pilot,Policy,Telemetry}Address should be sufficient.

Comment (linsun) [reply operator/data/profiles/remote.yaml]: Ok that SG!  thx!

File: manifests/base/templates/endpoints.yaml
Comment (linsun) [reply manifests/base/templates/endpoints.yaml]: isn't enabling pilot the only way to enable istiod?  That is what the remote profile has today.

Comment (ayj) [reply manifests/base/templates/endpoints.yaml]: istiod runs in each clsuter and we need the istiod-remote service to avoid conflicting the "real" istiod service in the cluster.

When istiod=false (separate services) pilot doesn't need to run in the remote cluster. The `istio-pilot` service in the remote cluster points to the pilot in the main cluster. This is how things worked in <= 1.4.

**Comment by ayj** [comment]
split-horizon is working now. Requests from each cluster are balanced roughly evenly across all clusters. I've tested with  HTTP echo tests so far. I still need to run more extensive traffic and load tests.

Cluster topology:
Master (network0)
Remote0, 2 (network1)
Remote1, (network2)

I'm still seeing ACK errors in the istiod and proxy logs. 

`
2020-03-07T19:00:17.066808Z	warn	ads	ADS:EDS: ACK ERROR 10.44.3.16:46908 router~10.44.3.16~istio-ingressgateway-5c4f66856f-fx46n.istio-system~istio-system.svc.cluster.local-6 Internal:Proto constraint validation failed (ClusterLoadAssignmentValidationError.Endpoints[i]: ["embedded message failed validation"] | caused by LocalityLbEndpointsValidationError.LbEndpoints[i]: ["embedded message failed validation"] | caused by LbEndpointValidationError.LoadBalancingWeight: ["value must be greater than or equal to " '\x01']): cluster_name: "outbound|15012||istiod.istio-system.svc.cluster.local"
`

It looks like this might be the cause which is similar to https://github.com/istio/istio/issues/18671 and https://github.com/istio/istio/issues/20087.

`
    load_balancing_weight {
    }
    1: 2           
  }
  load_balancing_weight {
    value: 4
  }
`

Full logs [here](https://gist.github.com/ayj/5227f9e5ea8db6bb62d8b78769d0d5f7)

**Comment by ayj** [comment]
@howardjohn @rshriram @costinm, any ideas on debugging the ACR error with zero lbEndpoints? I added extra logging to see where we might be setting weights to zero (or more likely not setting them at all). I didn't find anything. This is easily reproducible.

**Comment by howardjohn** [comment]
Shriram ran into this as well, not sure we found the root cause?

**Comment by kyessenov** [comment]
That `1:2` looks like proto mis-match due to field re-numbering.

**Comment by ayj** [comment]
The LB weights were always greater than zero immediately prior to sending. The issue didn't occur the endpoints proto cloned before calling stream.Send(). 

I instrumented pilot to run in-cluster as unit test with the race detector on. It found a data race in the eds path where the endpoint LB weights are updated/read. Looks like endpoints from the cached endpoint shared are reused across connections / pushes. I'm not sure how this is suppose to work for multiple networks where weights for the same service/endpoints could differ.

https://github.com/istio/istio/blob/ee52425661349e8a9b6973942001ca04c46c6b6a/pilot/pkg/proxy/envoy/v2/ep_filters.go#L58

https://github.com/istio/istio/blob/ee52425661349e8a9b6973942001ca04c46c6b6a/pilot/pkg/proxy/envoy/v2/eds.go#L876

Full race backtrace - https://gist.github.com/ayj/0bf0d1b8a349f90aaaeafd81b06863a5

**Comment by howardjohn** [comment]
> I instrumented pilot to run in-cluster as unit test with the race detector on. 

Not sure how much work this involved but you can do `go build -race` as well by the way I think

**Review by linsun — APPROVED** [review]
@ayj LG!  can we get this merged first?  I would like to try multi-network scenario too.

File: manifests/gateways/istio-egress/templates/deployment.yaml
Comment (howardjohn) [inline manifests/gateways/istio-egress/templates/deployment.yaml]: This doesn't do anything in master, it is reading from mesh config directly. 1.5 its needed though

**Comment by ayj** [comment]
Removed the hold. The bad endpoint LB weight doesn't seem to be introduced by this change, though it does increase the likelihood with multiple networks.

**Comment by ayj** [comment]
> Not sure how much work this involved but you can do go build -race as well by the way I think

Didn't know that existed - thanks.

File: manifests/base/templates/endpoints.yaml
Comment (howardjohn) [reply manifests/base/templates/endpoints.yaml]: We probably don't need to support the istiod=false case for multicluster

File: manifests/base/templates/services.yaml
Comment (howardjohn) [reply manifests/base/templates/services.yaml]: A bit confused why this is headless but istiod-remote is not? In the description you said istio-remote IS headless

File: manifests/istio-control/istio-discovery/templates/configmap.yaml
Comment (howardjohn) [inline manifests/istio-control/istio-discovery/templates/configmap.yaml]: defining variables like this doesn't actually work inside if statements. this is fixed on 1.5 but somehow broken on master, I have a fix in https://github.com/istio/istio/pull/21788/files

File: operator/data/profiles/default.yaml
Comment (howardjohn) [inline operator/data/profiles/default.yaml]: should this be `grpc-istiod`? Or tcp

File: manifests/base/templates/endpoints.yaml
Comment (ayj) [reply manifests/base/templates/endpoints.yaml]: I agree for not needing this in master. What about 1.5?

File: manifests/base/templates/services.yaml
Comment (linsun) [reply manifests/base/templates/services.yaml]: istiod remote is also headless because it doesn't have any selector.

File: manifests/istio-control/istio-discovery/templates/configmap.yaml
Comment (ayj) [reply manifests/istio-control/istio-discovery/templates/configmap.yaml]: Yup. Looks like your PR and my change below are similar except for the user of the DNS name instead of IP address.

Comment (linsun) [reply manifests/istio-control/istio-discovery/templates/configmap.yaml]: interesting, thanks!

File: operator/data/profiles/default.yaml
Comment (linsun) [reply operator/data/profiles/default.yaml]: I assume tcp, it is used everywhere for port 15012

**Review by esnible — APPROVED** [review]

File: manifests/base/templates/services.yaml
Comment (ayj) [reply manifests/base/templates/services.yaml]: Its missing `clusterIP: None`. I'll add it.

File: operator/data/profiles/default.yaml
Comment (ayj) [reply operator/data/profiles/default.yaml]: Changed to `tcp-istiod`

File: pilot/pkg/serviceregistry/aggregate/controller.go
Comment (rshriram) [inline pilot/pkg/serviceregistry/aggregate/controller.go]: why change this?

Comment (ayj) [reply pilot/pkg/serviceregistry/aggregate/controller.go]: Leftover from debugging. I'll revert.

File: pilot/cmd/pilot-discovery/main.go
Comment (ayj) [reply pilot/cmd/pilot-discovery/main.go]: It wasn't the same. `metav1.NamespaceAll` and the default ClusterRegistriesNamespace is `istio-system`. It makes sense to central the remote secrets and we don't need or want access to other namespace secrets.

File: pilot/pkg/bootstrap/certcontroller.go
Comment (ayj) [reply pilot/pkg/bootstrap/certcontroller.go]: Good point. We can skip adding the altName if it matches the hostname.

File: pilot/cmd/pilot-discovery/main.go
Comment (hzxuzhonghu) [reply pilot/cmd/pilot-discovery/main.go]: By a closer look:
1. serverArgs.Config.ClusterRegistriesNamespace is first parsed from cmd line, 
2. and then in `NewPilotArgs` it is set to `p.Namespace` or `constants.IstioSystemNamespace` accordingly. 

So it seems here the flag can be removed

**Comment by ayj** [comment]
/retest

leaderelection test is flakey

**Comment by howardjohn** [comment]
yeah it's a real bug in k8s, we cannot really avoid it. I may just skip the
test

File: pilot/cmd/pilot-discovery/main.go
Comment (ayj) [reply pilot/cmd/pilot-discovery/main.go]: Flags are parsed after NewPilotArgs() is call in the init() function. We don't use the flag in our manifests. There are probably a lot of other flags we could remove too. I'd rather leave that to a separate PR that doesn't get backported to 1.5

File: manifests/base/templates/services.yaml
Comment (sdake) [inline manifests/base/templates/services.yaml]: This may be a problem port - even in RFC1918 networks. Can you test without it?

File: manifests/base/templates/services.yaml
Comment (sdake) [inline manifests/base/templates/services.yaml]: This may be a problem port - even in RFC1918 networks. Can you test without it?

File: manifests/base/templates/services.yaml
Comment (sdake) [inline manifests/base/templates/services.yaml]: This may be a problem port - even in RFC1918 networks. Can you test without it?

File: manifests/istio-control/istio-discovery/templates/service.yaml
Comment (sdake) [inline manifests/istio-control/istio-discovery/templates/service.yaml]: extra CR.

File: manifests/base/templates/endpoints.yaml
Comment (sdake) [reply manifests/base/templates/endpoints.yaml]: @ayj I'd recommend sorting out what this should look like for `1.5` in master. Then follow up with another PR after we release 1.5.1.

File: manifests/base/templates/endpoints.yaml
Comment (sdake) [inline manifests/base/templates/endpoints.yaml]: This may be a problem port - can you test without in a pre istiod/post istiod world? (1.4 + 1.5 two cluster setup?)

File: manifests/base/templates/endpoints.yaml
Comment (sdake) [inline manifests/base/templates/endpoints.yaml]: This may be a problem port - can you test without in a pre istiod/post istiod world? (1.4 + 1.5 two cluster setup?)

File: manifests/base/templates/endpoints.yaml
Comment (sdake) [inline manifests/base/templates/endpoints.yaml]: This may be a problem port - can you test without in a pre istiod/post istiod world? (1.4 + 1.5 two cluster setup?)

File: manifests/base/templates/endpoints.yaml
Comment (sdake) [inline manifests/base/templates/endpoints.yaml]: 👍

File: manifests/base/templates/services.yaml
Comment (sdake) [reply manifests/base/templates/services.yaml]: @ayj istio-remote in 1.4 should have had `clusterIP: None` already - as the implementation was based upon a headless service. Do you mean that you missed that in this PR, or instead, that 1.4 has an issue on this point?

File: manifests/istio-control/istio-discovery/files/gen-istio.yaml
Comment (sdake) [inline manifests/istio-control/istio-discovery/files/gen-istio.yaml]: unneeded CR.

**Review by sdake — COMMENTED** [review]
Jason,

Really nice work. I left a few comments, some minor. I'd like you to take a look at port removal on istio-remote endpoint. When this work was originally implemented many ports were sloppily added for Pilot, but in the final implementation - 15011 was the only required port.

Cheers
-steve

File: manifests/base/templates/endpoints.yaml
Comment (ayj) [reply manifests/base/templates/endpoints.yaml]: Master doesn't have the legacy helm charts. We'll need to manually cherry-pick this PR anyway.

File: manifests/base/templates/services.yaml
Comment (ayj) [reply manifests/base/templates/services.yaml]: Refresh. The istiod-remote service has `ClusterIP: None` now. I forgot to add it in the initial commit.

**Comment by linsun** [comment]
any idea why this is closed @howardjohn @ayj ?

**Comment by linsun** [comment]
@ayj I pulled this PR locally and tested with master.  I'm getting this error during install.  This relates to a point I tried to make earlier that we are creating istio-pilot service as part of istiod enabled in remote cluster which will create the istio-pilot endpoint using the selector indirectly, so when we ask to specifically to create the istio-pilot endpoint it will fail.

```
Error from server (Invalid): error when applying patch:
{"metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"v1\",\"kind\":\"Service\",\"metadata\":{\"annotations\":{},\"labels\":{\"app\":\"pilot\",\"istio\":\"pilot\",\"release\":\"istio\"},\"name\":\"istio-pilot\",\"namespace\":\"istio-system\"},\"spec\":{\"ports\":[{\"name\":\"grpc-xds\",\"port\":15010},{\"name\":\"https-xds\",\"port\":15011},{\"name\":\"https-dns\",\"port\":15012},{\"name\":\"http-legacy-discovery\",\"port\":8080},{\"name\":\"http-monitoring\",\"port\":15014},{\"name\":\"https-webhook\",\"port\":443,\"targetPort\":15017}],\"selector\":{\"istio\":\"pilot\"}}}\n"},"labels":{"app":"pilot","istio":"pilot","release":"istio"}},"spec":{"$setElementOrder/ports":[{"port":15010},{"port":15011},{"port":15012},{"port":8080},{"port":15014},{"port":443}],"clusterIP":null,"ports":[{"name":"https-dns","port":15012},{"name":"https-webhook","port":443,"targetPort":15017}],"selector":{"istio":"pilot"}}}
to:
Resource: "/v1, Resource=services", GroupVersionKind: "/v1, Kind=Service"
Name: "istio-pilot", Namespace: "istio-system"
Object: &{map["apiVersion":"v1" "kind":"Service" "metadata":map["annotations":map["kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"v1\",\"kind\":\"Service\",\"metadata\":{\"annotations\":{},\"name\":\"istio-pilot\",\"namespace\":\"istio-system\"},\"spec\":{\"clusterIP\":\"None\",\"ports\":[{\"name\":\"grpc-xds\",\"port\":15010},{\"name\":\"https-xds\",\"port\":15011},{\"name\":\"http-legacy-discovery\",\"port\":8080},{\"name\":\"tcp-istiod\",\"port\":15012},{\"name\":\"http-monitoring\",\"port\":15014}]}}\n"] "creationTimestamp":"2020-03-10T15:01:10Z" "name":"istio-pilot" "namespace":"istio-system" "resourceVersion":"48395868" "selfLink":"/api/v1/namespaces/istio-system/services/istio-pilot" "uid":"d8233156-0939-41d2-96b7-1130bcb3a216"] "spec":map["clusterIP":"None" "ports":[map["name":"grpc-xds" "port":'\u3aa2' "protocol":"TCP" "targetPort":'\u3aa2'] map["name":"https-xds" "port":'\u3aa3' "protocol":"TCP" "targetPort":'\u3aa3'] map["name":"http-legacy-discovery" "port":'\u1f90' "protocol":"TCP" "targetPort":'\u1f90'] map["name":"tcp-istiod" "port":'\u3aa4' "protocol":"TCP" "targetPort":'\u3aa4'] map["name":"http-monitoring" "port":'\u3aa6' "protocol":"TCP" "targetPort":'\u3aa6']] "sessionAffinity":"None" "type":"ClusterIP"] "status":map["loadBalancer":map[]]]}
for: "test-remote.yaml": Service "istio-pilot" is invalid: spec.clusterIP: Invalid value: "": field is immutable
Error from server: error when creating "test-remote.yaml": context deadline exceeded
```

File: manifests/gateways/istio-egress/templates/deployment.yaml
Comment (howardjohn) [reply manifests/gateways/istio-egress/templates/deployment.yaml]: can you remove after this merges?

File: pilot/pkg/serviceregistry/kube/controller/controller.go
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/controller.go]: nit: I think this should be a warning. Its not an internal error, its just sometimes not possible from the metadata provided by the client

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/controller.go]: Actually I didn't notice you aren't logging this. We should not return an error here, I think. Was there a reason you think we should?

**Review by howardjohn — CHANGES_REQUESTED** [review]
Everything lgtm except the minor error comment

File: pilot/pkg/serviceregistry/kube/controller/controller.go
Comment (ayj) [reply pilot/pkg/serviceregistry/kube/controller/controller.go]: I went back and forth on changing this. I was seeing a lot of these messages in the logs. It didn't seem like an error/warning when using multiple control planes. The check on line 541 will fail because Cluster A's registry won't have any pods for Cluster B. And getProxyServiceInstancesFromMetadata will fail because the remote proxy's cluster_id doesn't match the cluster's ID. It should still be printed but as a debug statement by the aggregate.

Some other alternatives:
* modify getProxyServiceInstancesFromMetadata to not return an error when cluster ID's didn't match. 
* modify the aggregate controller to only call GetProxyServiceInstances() on the controller whose cluster_id matches the target proxy.

File: manifests/gateways/istio-egress/templates/deployment.yaml
Comment (ayj) [reply manifests/gateways/istio-egress/templates/deployment.yaml]: Yup - https://github.com/istio/istio/issues/22050

File: pilot/pkg/serviceregistry/kube/controller/controller.go
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/controller.go]: I am fine with making changes so this logs nothing in multicluster case - that makes complete sense. but if we return an error here, won't the aggregate controller bubble that error up and ultimately we will return an error? So it will always be an error for MC case? I am surprised this works at all, am I missing some step?

Comment (ayj) [reply pilot/pkg/serviceregistry/kube/controller/controller.go]: AFAICT the aggregate controller will log the error but not bubble it up if another controller returned one or more instances (see 
https://github.com/istio/istio/blob/f8464b5eb8f06ee67cf4b1074d33b27610dd9f08/pilot/pkg/serviceregistry/aggregate/controller.go#L266-L270). However, if there is no instances the error eventually bubbles up to initConnection which fails the connection and is not something we want.

How about the following? cc @nmittler 

```diff
diff --git a/pilot/pkg/serviceregistry/aggregate/controller.go b/pilot/pkg/serviceregistry/aggregate/controller.go
index 1741a762e..6964364a4 100644
--- a/pilot/pkg/serviceregistry/aggregate/controller.go
+++ b/pilot/pkg/serviceregistry/aggregate/controller.go
@@ -228,6 +228,9 @@ func (c *Controller) GetProxyServiceInstances(node *model.Proxy) ([]*model.Servi
        // It doesn't make sense for a single proxy to be found in more than one registry.
        // TODO: if otherwise, warning or else what to do about it.
        for _, r := range c.GetRegistries() {
+               if r.Cluster() != "" && node.Metadata != nil && node.Metadata.ClusterID != "" && node.Metadata.ClusterID != r.Cluster() {
+                       continue
+               }
                instances, err := r.GetProxyServiceInstances(node)
                if err != nil {
                        errs = multierror.Append(errs, err)
```

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/controller.go]: Ah thanks... I looked right at that code and missed it. I think that makes sense, probably means things up a bit in general

Comment (ayj) [reply pilot/pkg/serviceregistry/kube/controller/controller.go]: Updated to filter in the aggregate controller. We need to special case the default cluster which is always name `Kubernetes`. I started to clean things up so the default registry uses the provided ClusterName but decided defer that for another PR. This is large enough as-is.

File: pilot/pkg/serviceregistry/aggregate/controller.go
Comment (howardjohn) [inline pilot/pkg/serviceregistry/aggregate/controller.go]: So if `r.Cluster == Kubernetes` and `node.Cluster == MyOtherCluster`, this will NOT be skipped? Isn't that what we were trying to avoid?

File: pilot/pkg/serviceregistry/aggregate/controller.go
Comment (howardjohn) [inline pilot/pkg/serviceregistry/aggregate/controller.go]: Why this change? This means we can only get endpoints from one registry? Which will break all the other registries like ServiceEntry?

Comment (ayj) [reply pilot/pkg/serviceregistry/aggregate/controller.go]: See the for loop comment - "It doesn't make sense for a single proxy to be found in more than one registry".  `out` only ever has at most instances from one registry. The code breaks out of the loop two lines below on line 251.

Comment (howardjohn) [reply pilot/pkg/serviceregistry/aggregate/controller.go]: I don't think that comment is accurate? I am surprised all tests pass, I think you should be able to have ServiceEntry + Kube.

@rshriram ?

File: pilot/pkg/serviceregistry/aggregate/controller.go
Comment (ayj) [reply pilot/pkg/serviceregistry/aggregate/controller.go]: Yes, that is one of the cases we were trying to avoid. ClusterID is overloaded to mean both the registry type (k8s, consul) and the name (which cluster). I didn't want to fix that issue holistically in this PR (see / https://github.com/istio/istio/issues/22093). On further thought though we could filter 100% of the cases if we had some special handling above that is less intrusive (see skipSearchingRegistryForProxy above).

File: pilot/pkg/serviceregistry/aggregate/controller.go
Comment (ayj) [reply pilot/pkg/serviceregistry/aggregate/controller.go]: I've removed the change from this PR. We can address it in a follow-up.

**Review by howardjohn — APPROVED** [review]

**Comment by linsun** [comment]
@ayj any plan to back port this to 1.5?

**Comment by ayj** [comment]
Yes. PR is in progress. I'm testing the upgrade/downgrade cases. 1.4<=>1.5 with istiodctl and istiod and non-istiod variants are passing.

**Pull request merged**
